### PR TITLE
Is valid static caching thread safety

### PR DESCRIPTION
**Pull request by michaelcfanning**
Our static analysis validation phase contains a per-rule caching mechanism that is intended to help prevent multiple scan results for the same unique secret firing against the same file. This pattern can occur due to poorly behaved regexes and so the rules, in this case, report the first occurrence only (and drop the remainder).

The backing data structure, a HashSet<string> isn't thread-safe, and so in complex analysis scenarios we see exceptions in this code path: an 'InvalidOperationException' (in cases where HashSet.Contains provoked the issue) or 'IndexOutOfRange' (in cases where HashSet,Add provoked the problem).

The solution is to convert to using a ConcurrentDictionary instance (as .NET has no built-in thread-safe HashSet collection). This is a bit idiosyncratic, but there's no strong driving need to create a custom type (or accept a new NuGet 3rd party dependency) to get a cleaner solution. Simply using ConcurrentDictionary as a set of key is sufficient for now.

To fix this bug, I first authored a multithreaded test to provoke the issue, which now passes. This test adds ~500 ms of execution on my machine (and fails ~50% of the time without the fix).

@EasyRhinoMSFT @marmegh @shaopeng-gh @yongyan-gh @eddynaka

### CODE REVIEW

File: Src/ReleaseHistory.md [hunk 1]
```diff
@@ -18,6 +18,8 @@
 
 ## *Unreleased*
 
+- BUG: Resolve `InvalidOperationException` and `IndexOutOfRange` exceptions in `StaticValidatorBase.IsValidStatic` due to unsafe use of HashSet<string> class.
+  [#595](https://github.com/microsoft/sarif-pattern-matcher/pull/585)
```
Comment (shaopeng-gh) [inline Src/ReleaseHistory.md:22]: >585 [](http://example.com/codeflow?start=65&length=3)

wrong number

File: Src/Sarif.PatternMatcher.Sdk/IDictionary.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk
+{
+    public interface IDictionary<T>
```
Comment (shaopeng-gh) [inline Src/Sarif.PatternMatcher.Sdk/IDictionary.cs:6]: >IDictionary [](http://example.com/codeflow?start=21&length=11)

is this file needed

File: Src/Sarif.PatternMatcher.Sdk/StaticValidatorBase.cs [hunk 3]
```diff
@@ -23,13 +23,13 @@ public IEnumerable<ValidationResult> IsValidStatic(IDictionary<string, FlexMatch
...
 
-                if (PerFileFingerprintCache.Contains(key))
+                if (PerFileFingerprintCache.ContainsKey(key))
                 {
                     validationResult.ValidationState = ValidationState.NoMatch;
                     continue;
                 }
 
-                PerFileFingerprintCache.Add(key);
+                PerFileFingerprintCache.Add(key, (byte)0);
```
Comment (shaopeng-gh) [inline Src/Sarif.PatternMatcher.Sdk/StaticValidatorBase.cs:32]: >(byte)0 [](http://example.com/codeflow?start=49&length=7)

also can use default(byte)

File: Src/Test.UnitTests.Sarif.PatternMatcher.Sdk/StaticValidatorBaseTests.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            {
+                Interlocked.Increment(ref id);
+
+                for (int iterations = 0; iterations < 100; iterations++)
+                {
+                    var groups = new Dictionary<string, FlexMatch>
+                {
+                    { "scanTargetFullPath", new FlexMatch { Value = Guid.NewGuid().ToString()} },
+                    { "0", new FlexMatch { Value = Guid.NewGuid().ToString() } }
+                };
```
Comment (shaopeng-gh) [inline Src/Test.UnitTests.Sarif.PatternMatcher.Sdk/StaticValidatorBaseTests.cs:46]: is the alignment correct here or need to indent

**Review by shaopeng-gh — APPROVED** [review]
:shipit:

File: Src/ReleaseHistory.md [hunk 5]
```diff
@@ -18,6 +18,8 @@
 
 ## *Unreleased*
```
Comment (shaopeng-gh) [inline Src/ReleaseHistory.md:19]: >Unreleased [](http://example.com/codeflow?start=4&length=10)

dotnet format task and build task is in failed status

Comment (marmegh) [reply Src/ReleaseHistory.md:19]: I have another PR open to fix that. It's caused by an ADO update.
